### PR TITLE
[8.0][ADD] web_widget_image_download

### PR DESCRIPTION
**Pull request by kutyel**
# Web Widget - Image Download

This module was written to extend the functionality of the widget image and allows to download it.
# Usage

To use this module, you need to:
1. Go to the section `Contacts`.
2. Click on a contact.
3. Edit the contact.
4. Click the `Download` button (between `Edit` and `Clear`).

@Yajo @Tecnativa

### CODE REVIEW

**Comment by pedrobaeza** [comment]
You can put a better icon using a screenshot of the usual icon size (128x128) showing the download button. If you let the default icon, there's no need of keeping the SVG file, as it's already in the template.

**Comment by pedrobaeza** [comment]
CLA ready

File: web_widget_image_download/README.rst [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/162/8.0
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/{project_repo}/issues>`_. In case of trouble, please
```
Comment (yajo) [inline web_widget_image_download/README.rst]: Replace with `web`

File: web_widget_image_download/README.rst [hunk 2]
```diff
@@ -0,0 +1,59 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
```
Comment (yajo) [inline web_widget_image_download/README.rst]: LGPL

File: web_widget_image_download/README.rst [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+===========================
+Web Widget - Image Download
+===========================
+
+This module was written to extend the functionality of the widget image and allow the download of the contact avatar.
```
Comment (yajo) [inline web_widget_image_download/README.rst]: s/widget image/image widget/

... and not all images are contact avatars. :wink:

File: web_widget_image_download/static/description/icon.svg [hunk 4]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (yajo) [inline web_widget_image_download/static/description/icon.svg]: If you use the default OCA icon, you don't need to include its SVG.

File: web_widget_image_download/static/src/css/web_widget_image_download.css [hunk 5]
```diff
@@ -0,0 +1,6 @@
+/* Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). */
+
+.oe_form_binary_file_download {
+  color: #eee !important;
```
Comment (yajo) [inline web_widget_image_download/static/src/css/web_widget_image_download.css]: If `!important` is not required, it's best to avoid it.

File: web_widget_image_download/static/src/css/web_widget_image_download.css [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+}
```
Comment (yajo) [inline web_widget_image_download/static/src/css/web_widget_image_download.css]: Always end files with newline.

File: web_widget_image_download/static/src/js/web_widget_image_download.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+ */
```
Comment (yajo) [inline web_widget_image_download/static/src/js/web_widget_image_download.js:4]: You can save a couple of LOC removing those dangling comment endings, [like in the template](https://github.com/OCA/maintainer-tools/blob/master/template/module/static/src/js/module_name.js#L1-L2). Not a blocker though.

File: web_widget_image_download/static/src/css/web_widget_image_download.css (hunk 5 above)
Comment (kutyel) [inline web_widget_image_download/static/src/css/web_widget_image_download.css]: @Yajo you are right, I didn't want to include it, but there was another class from odoo that override my widget custom css, so I had to do it 😕

File: web_widget_image_download/static/src/js/web_widget_image_download.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+openerp.web_widget_image_download = function(instance) {
+    'use strict';
+
+    instance.web.form.web_widget_image_download = instance.web.form.FieldBinaryImage.include({
+        render_value() {
```
Comment (yajo) [inline web_widget_image_download/static/src/js/web_widget_image_download.js:9]: What's the current browser compatibility for this?

File: web_widget_image_download/static/src/xml/web_widget_image_download.xml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<template>
+    <t t-extend="FieldBinaryImage">
+        <t t-jquery=".oe_form_binary_file_edit" t-operation="after">
+            <a class="fa fa-download fa-1g col-md-4 oe_form_binary_file_download" title="Download" download="avatar.png"></a>
```
Comment (pedrobaeza) [inline web_widget_image_download/static/src/xml/web_widget_image_download.xml]: Call it `image.png`, not `avatar.png`

File: web_widget_image_download/static/src/js/web_widget_image_download.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+            this._super();
```
Comment (yajo) [inline web_widget_image_download/static/src/js/web_widget_image_download.js:10]: Shouldn't you return this?

File: web_widget_image_download/README.rst (hunk 2 above)
Comment (kutyel) [inline web_widget_image_download/README.rst]: @Yajo you want me to change all the licenses to LGPL? 😨

File: web_widget_image_download/static/src/js/web_widget_image_download.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+            // remove `col-md-offset-5` from following trash icon (improves visibility)
+            this.$el.find('.oe_form_binary_file_clear').removeClass('col-md-offset-5');
```
Comment (yajo) [inline web_widget_image_download/static/src/js/web_widget_image_download.js]: View replacements should be done through QWeb if possible.

File: web_widget_image_download/static/src/xml/web_widget_image_download.xml (hunk 9 above)
Comment (pedrobaeza) [inline web_widget_image_download/static/src/xml/web_widget_image_download.xml]: You should translate this text. @Yajo, can you explain to him the method to translate templates?

File: web_widget_image_download/README.rst (hunk 2 above)
Comment (antespi) [inline web_widget_image_download/README.rst]: All OCA addons are AGPL-3, no LGPL
Why LGPL, can you explain it?

File: web_widget_image_download/static/src/xml/web_widget_image_download.xml (hunk 9 above)
Comment (yajo) [inline web_widget_image_download/static/src/xml/web_widget_image_download.xml]: What's this `download` attribute?

File: web_widget_image_download/README.rst (hunk 2 above)
Comment (yajo) [inline web_widget_image_download/README.rst]: If you ever want to create an LGPL website addon (such as a theme) and you want to depend on this module, you'd be violating the license.

Comment (pedrobaeza) [inline web_widget_image_download/README.rst]: LGPL should be for website (frontend) modules, not web (backend) modules

File: web_widget_image_download/static/src/css/web_widget_image_download.css (hunk 5 above)
Comment (yajo) [inline web_widget_image_download/static/src/css/web_widget_image_download.css]: Can't you just use a more specific selector to get higher priority?

File: web_widget_image_download/README.rst (hunk 2 above)
Comment (pedrobaeza) [inline web_widget_image_download/README.rst]: Talking with @Yajo, he is right about the possibility of using this widget through https://github.com/OCA/website/tree/8.0/website_backend_views, and in this case, having an LGPL widget that can be reused in any frontend is a better option.

File: web_widget_image_download/static/src/xml/web_widget_image_download.xml (hunk 9 above)
Comment (kutyel) [inline web_widget_image_download/static/src/xml/web_widget_image_download.xml]: @Yajo its the [standard HTML5 download attribute](http://www.w3schools.com/tags/att_a_download.asp), required to download the image.

File: web_widget_image_download/static/src/js/web_widget_image_download.js (hunk 8 above)
Comment (kutyel) [inline web_widget_image_download/static/src/js/web_widget_image_download.js:9]: Quite [good](http://kangax.github.io/compat-table/es6/#test-object_literal_extensions), is the new _shorthand method_ for object literals. 😄

**Comment by kutyel** [comment]
Thank you all for the feedback @Tecnativa! :+1:

File: web_widget_image_download/static/src/js/web_widget_image_download.js (hunk 10 above)
Comment (kutyel) [inline web_widget_image_download/static/src/js/web_widget_image_download.js:10]: @Yajo I don't think so, the parent widget doesn't and it does not use it either.

File: web_widget_image_download/static/src/js/web_widget_image_download.js (hunk 11 above)
Comment (antespi) [inline web_widget_image_download/static/src/js/web_widget_image_download.js]: QWeb Engine in JS side doesn't support attributes position. In this case, we must use replace. And I don't like replaces because they could break inheritance. In this case I think that JQuery manipulation is the safest way.

**Comment by rafaelbn** [comment]
Tested un runbot :+1:

**Comment by antoniocanovas** [comment]
Tested un runbot 👍

File: web_widget_image_download/README.rst (hunk 2 above)
Comment (legalsylvain) [inline web_widget_image_download/README.rst]: Hi, 
I understand that for compatibility reasons, it is better to have LGPL, but contributors has the right to share their work with OCA in AGPL.

Comment (pedrobaeza) [inline web_widget_image_download/README.rst]: @legalsylvain, he is a new guy in our team (Tecnativa), so we are deciding publicly which license to use.

Comment (legalsylvain) [inline web_widget_image_download/README.rst]: OK, I did'nt know.

:+1:

File: web_widget_image_download/README.rst [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/web/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed feedback.
+
+Credits
+=======
+
+Images
```
Comment (kutyel) [inline web_widget_image_download/README.rst:36]: Should have removed this section, the image was mine.

File: web_widget_image_download/__openerp__.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+{
+    "name": "Web Widget - Image Download",
+    "summary": "Allows to download the avatar for the contact",
```
Comment (kutyel) [inline web_widget_image_download/__openerp__.py:6]: This should be:

```
This module was written to extend the functionality of the widget image and allows to download it.
```

File: web_widget_image_download/views/assets.xml [hunk 14]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2016 Flavio Corpa <flavio.corpa@tecnativa.com>
+     License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl). -->
+
+<openerp>
```
Comment (kutyel) [inline web_widget_image_download/views/assets.xml:5]: I bet this will be `<odoo>` in v9.0?

File: web_widget_image_download/README.rst (hunk 12 above)
Comment (pedrobaeza) [inline web_widget_image_download/README.rst:36]: Yeah, but you can make another PR to amend it. There are 2 more things that I have thought about this weekend:
- If there's no image, is there the button to download?
- If the image is not png, the extension shouldn't be png (for example, jpg or gif). We should detect the type. Is there an easy way for that?

File: web_widget_image_download/views/assets.xml (hunk 14 above)
Comment (pedrobaeza) [inline web_widget_image_download/views/assets.xml:5]: Yeah, and you can remove `<data>` also in that version

File: web_widget_image_download/README.rst (hunk 12 above)
Comment (kutyel) [inline web_widget_image_download/README.rst:36]: Yes, @Yajo told me this weekend that an improvement would be to use odoo's `t-attr-` tags to retrieve the file name (and format) from the server and then add it dynamically.

File: web_widget_image_download/__openerp__.py (hunk 13 above)
Comment (yajo) [inline web_widget_image_download/__openerp__.py:6]: Actually the "this module was written to ..." part is more for the README, and lately just removed (all are written modules after all). However something like "Allows to download any image from its widget" seems better.

**Comment by yajo** [comment]
Congratulations @kutyel! :tada:
